go/vt/mysqlctl/azblobbackupstorage/azblob.go

```diff
@@ -0,0 +1,297 @@
+type AZBlobBackupStorage struct {
+}
+
+func (bs *AZBlobBackupStorage) containerURL() (*azblob.ContainerURL, error) {
+	credentials, err := azCredentials()
+	if err != nil {
+		return nil, err
+	}
+	u := azServiceURL(credentials).NewContainerURL(*containerName)
+	return &u, nil
+}
+
+// ListBackups implements BackupStorage.
+func (bs *AZBlobBackupStorage) ListBackups(ctx context.Context, dir string) ([]backupstorage.BackupHandle, error) {
+	log.Infof("ListBackups: [azblob] container: %v, prefix: %v", *containerName, dir)
+
+	containerURL, err := bs.containerURL()
+	if err != nil {
+		return nil, err
+	}
+
+	searchPrefix := objName(dir, "")
+
+	result := make([]backupstorage.BackupHandle, 0)
+	var subdirs []string
+
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		resp, err := containerURL.ListBlobsHierarchySegment(ctx, marker, delimiter, azblob.ListBlobsSegmentOptions{
+			Prefix:     searchPrefix,
+			MaxResults: 0,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		for _, item := range resp.Segment.BlobPrefixes {
+			subdir := strings.TrimPrefix(item.Name, searchPrefix)
+			subdir = strings.TrimSuffix(subdir, delimiter)
+			subdirs = append(subdirs, subdir)
+		}
+
+		marker = resp.NextMarker
+	}
+
+	for _, subdir := range subdirs {
+		result = append(result, &AZBlobBackupHandle{
+			bs:       bs,
+			dir:      dir,
+			name:     subdir,
+			readOnly: true,
+		})
+	}
+
+	return result, nil
+}
+
+// StartBackup implements BackupStorage.
+func (bs *AZBlobBackupStorage) StartBackup(ctx context.Context, dir, name string) (backupstorage.BackupHandle, error) {
+	return &AZBlobBackupHandle{
+		bs:       bs,
+		dir:      dir,
+		name:     name,
+		readOnly: false,
+	}, nil
+}
+
+// RemoveBackup implements BackupStorage.
+func (bs *AZBlobBackupStorage) RemoveBackup(ctx context.Context, dir, name string) error {
+	log.Infof("ListBackups: [azblob] container: %v, prefix: %v", *containerName, dir)
+
+	containerURL, err := bs.containerURL()
+	if err != nil {
+		return err
+	}
+
+	searchPrefix := objName(dir, name, "")
+
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		resp, err := containerURL.ListBlobsHierarchySegment(ctx, marker, delimiter, azblob.ListBlobsSegmentOptions{
+			Prefix:     searchPrefix,
+			MaxResults: 0,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		// Right now there is no batch delete so we must iterate over all the blobs to delete them one by one
+		// One day we will be able to use this https://docs.microsoft.com/en-us/rest/api/storageservices/blob-batch
+		// but currently it is listed as a preview and its not in the go API
+		for _, item := range resp.Segment.BlobItems {
+			_, err = containerURL.NewBlobURL(item.Name).Delete(ctx, azblob.DeleteSnapshotsOptionNone, azblob.BlobAccessConditions{})
+			if err != nil {
+				return err
+			}
+		}
+		marker = resp.NextMarker
+	}
+
+	return nil
+}
+
+// Close implements BackupStorage.
+func (bs *AZBlobBackupStorage) Close() error {
+	// This function is a No-op
```

Comment:
The S3 plugin at least unassigns the underlying client. If you see a reason to not do that (maybe adding the locks everywhere isn't worth it?), let's at least leave a comment explaining why we don't do it here.

Reply:
In this case there isn't really anything to unassigned. The BackupStorage structure doesn't actually contain anything. The credentials which are the only thing that are stored are static.

Reply:
Oh I see. I was looking at the wrong struct definition. Carry on.